microsoftgraph/msgraph-sdk-python-core
# Feat/retry handler

## Overview

Provides a configurable RetryHandler that handles the retrying of failed requests.

### Demo

Initializing a sample client with default middleware using the `HTTPClientFactory` and passing default configs to the `RetryHandler`.

```py
client = HTTPClientFactory().create_with_default_middleware(
        browser_credential, max_retries=5, retry_backoff_factor=0.1, retry_time_limit=60
    )
    result = client.get('/me/messages', scopes=['mail.read'])
    pprint(result.json())
```
Sending a request using the graph client with retry options for that specific request.

```py
client = GraphClient(credential=browser_credential)
    result = client.get(
        '/me/messages', scopes=['mail.read'], retry_on_status_codes=[429, 502, 503, 504]
    )
    pprint(result.json())
```
## Testing Instructions

* Pull the `feat/retry-handler` branch
* Instantiate a client and make a call to an endpoint that will return one of the retryable status codes (`429`, `503,` `504`)
```py
client = GraphClient(credential=my_credential)
resp = client.get('https://httpbin.org/status/429')

```
* Observe that the request was retried before the response is returned. This can be confirmed by observing the `retry_attempt` header in the request headers from the requests header
```py
print(resp.request.headers)
```

Fixes #8
